로또 1단계 미션 제출합니다

iimpala opened the pull request:
안녕하세요! 코드를 리팩토링하다가 제가 발견한 문제점과 질문사항들을 공유드립니다.
todo 관련해서 제가 잘못 생각하고 있는 부분이나 질문드린 내용을 포함해서 피드백 해주시면 감사하겠습니다😃

## TODO
- View 레이어가 도메인의 Prize를 알고있음 -> generatePrizeString을 컨트롤러로 이전하여 의존성을 분리할 계획
- 도메인 객체 내부에 toDto() 메소드 제거
- 응집성을 높이기 위해 WinningNumbers의 getPrize() 메소드를 Prize에 넣어 처리할까 생각중(가능한지는 확인해봐야 할 듯)
- Controller에서 NumberPool 대신 Strategy를 주입할 예정

## 질문
- PrizeCalculator와 Budget에서 TICKET_PRICE 상수를 중복해서 선언하는데 한 곳에서 관리하려면 어떤 방향으로 설계를 변경해야할지 피드백 부탁드립니다:)
- Controller의 start() 메소드에서 tickets, result에 assert not null 검사를 하라고 가이드가 나오는데 어떻게 하면 좀 더 괜찮아질까요?

adeldel commented on src/main/java/domain/LottoRank.java:
```diff
@@ -32,11 +31,10 @@
     }
 
     public static LottoRank calculateLottoRank(int count, boolean validBonusBall) {
```
이 메소드도 더이상 사용하지 않네요. 리팩토링 과정에서 불필요해진 메소드는 정리 부탁드립니다.

cyrojyro replied:
메인 코드에 없고 테스트 코드에만 덜렁 있는 코드네요.
직접 짠 저도 모르는 메서드를 찾아내시다니.. 대단하십니다 🆒

adeldel commented on src/main/java/domain/LottoRank.java:
```diff
@@ -32,11 +31,10 @@
+                .orElse(LottoRank.NONE);
     }
 
     public boolean checkRank(int count, boolean validBonusBall) {
```
`validBonusBall  -> hasBonusBall` 이 더 자연스러운거 같아요.
그리고 getter 메소드를 사용하지 않고 바로 enum 상수값으로 비교하면 되지 않을까요? 
이렇게 바꾸면 외부에서 쓰지 않는다면 getter 메소드는 지워도 될것 같습니다.
```
        if (bonusStateBallState == BonusState.IRRELEVANT) {
            return sameBallNumber == count;
        }

adeldel commented on src/main/java/domain/LottoRank.java:
```diff
@@ -61,9 +59,9 @@ public BonusState getBonusBallState() {
     }
 
-    public boolean checkRank(Lotto lotto, Lotto winningLotto) {
+    public boolean checkRank(Lotto lotto, WinningLotto winningLotto) {
```
위에 존재하는 `checkRank`와 중복되어 보입니다. 호출하는 곳에서 위 메소드로 변경하면 이 메소드도 정리할 수 있지 않을까요?

adeldel commented on src/main/java/domain/LottoRank.java:
```diff
@@ -61,9 +59,9 @@ public BonusState getBonusBallState() {
+                lotto.hasBonusBall(winningLotto.getBonusBall()));
     }
 
     public BigInteger getReward() {
```
리워드값이 크지 않다면 BigInteger로 형변환하지 않고 그대로 리턴해도 써도 되지 않을까 싶습니다. 
전반적으로 BigInteger로 사용하는 곳곳이 보이는데, 이렇게 형변환해서 사용하는 이유가 있을까요? 🤔

cyrojyro replied:
엄청나게 낮은 확률이지만, 1등이 2개 당첨되면 40억으로 int범위를 넘어서기 때문입니다. long을 쓰는 방법도 있지만, 한계가 거의 없는 BigInteger를 사용해 봤습니다.

cyrojyro replied:
금액 말고 다른 부분에는 굳이 BigInteger가 필요 없을 것 같아 int로 대체하는 방향으로 수정합니다.

adeldel commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.List;
+
+public class WinningLotto extends Lotto {
```
상속 👍

adeldel reviewed (changes requested):
안녕하세요 잭 😃 
두번째 로또미션도 잘 보았어요. 요구사항에 대한 구현이 잘 이루어 진 것으로 보입니다. 
이전 리뷰반영도 잘 해주셔서 감사합니다.  구조상 크게 변경하지 않아도 되지만, 불필요해 보이는 로직들이 곳곳에 있어서 쪼금 더 정리하면 좋을 것 같아 재요청드립니다.
수고 많으셨습니다.  👍

cyrojyro commented on src/main/java/domain/Lottos.java:
```diff
@@ -18,19 +18,26 @@ public Lottos(List<Lotto> lottos) {
+        List<Lotto> lottos = new ArrayList<>();
+        lottos.addAll(lottos1.getLottos());
+        lottos.addAll(lottos2.getLottos());
+        this.lottos = lottos;
```
맞네요. flatmap도 있었네요...

adeldel commented on src/test/java/domain/LottosTest.java:
```diff
@@ -32,31 +31,25 @@ public void createLottos() {
+        assertThat(lottos.calculateNumberOfRank(LottoRank.THIRD, winningLotto)).isEqualTo(2);
+        assertThat(lottos.calculateNumberOfRank(LottoRank.FOURTH, winningLotto)).isEqualTo(0);
+        assertThat(lottos.calculateNumberOfRank(LottoRank.FIFTH, winningLotto)).isEqualTo(0);
+        assertThat(lottos.calculateNumberOfRank(LottoRank.NONE, winningLotto)).isEqualTo(1);
```
👍

adeldel commented on src/test/java/domain/LottosTest.java:
```diff
@@ -32,31 +31,25 @@ public void createLottos() {
+        trueRankNumbers.put(LottoRank.THIRD, 2);
+        trueRankNumbers.put(LottoRank.FOURTH, 0);
+        trueRankNumbers.put(LottoRank.FIFTH, 0);
+        trueRankNumbers.put(LottoRank.NONE, 1);
```
👍

adeldel commented on src/main/java/domain/LottoRank.java:
```diff
@@ -17,36 +17,32 @@
+        return this.getSameBallNumber() == count && !hasBonusBall;
     }
 
     private int getSameBallNumber() {
```
요것도 위 checkRank 로직에서 수정하면 삭제할 수 있겠네요.

adeldel commented on src/main/java/domain/Lottos.java:
```diff
@@ -18,19 +18,26 @@ public Lottos(List<Lotto> lottos) {
+    }
+
     public List<Lotto> getLottos() {
         return new ArrayList<>(lottos);
```
요기 수정은 스킵하셨네요 🙈

adeldel commented on src/main/java/domain/Lotto.java:
```diff
@@ -10,29 +8,19 @@
 
-    public Lotto(List<LottoBall> balls, LottoBall bonusBall) {
-        this(new LottoBalls(balls), bonusBall);
+    public Lotto(List<LottoBall> balls) {
```
이 생성자는 유닛테스트에서만 사용하는 생성자입니다. 되도록이면 테스트에서만 사용하는 생성자는  자양해주세요.
유닛테스트에서 LottoBalls을 생성하여 넘겨주는 방식으로  변경가능해 보입니다.

cyrojyro replied:
테스트에서만 사용하는 코드들을 종종 메인 코드에 사용하였는데, 이부분에 대해 별도로 분리하는 습관을 들여야겠습니다.

adeldel reviewed (approved):
안녕하세요! 두번째 미션도 반영을 잘 해주셨고, 기능 요구사항도 잘 구현해주셨습니다.
로또번호를 미리 생성하여 재사용한 점도 칭찬드립니다. 👍 
소소하게 댓글을 달아드렸으나, 다음 미션시 참고하시면 좋을 것 같아요.
수고 많으셨습니다. 🎉

cyrojyro commented on src/main/java/domain/Lottos.java:
```diff
@@ -18,20 +21,26 @@ public Lottos(List<Lotto> lottos) {
     }
 
+    public Lottos(Lottos lottos1, Lottos lottos2) {
+        lottos = Stream.of(lottos1.getLottos(), lottos2.getLottos())
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
```
구버전을 보신듯 합니다

cyrojyro commented:
그동안 코드리뷰 해주셔서 감사드립니다!

chanhalee committed:
docs(README) Todo List 작성

Todo List 작성

chanhalee committed:
feat(Budget) 구매 금액에 대한 요구사항 Test 작성 및 구현

구매 금액에 대한 요구사항 Test 작성 및 구현

chanhalee committed:
feat(Ball) 로또 번호에 대한 요구사항 test 작성 및 구현

로또 번호에 대한 요구사항 test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 정수여야 한다.

chanhalee committed:
feat(Ticket) 티켓에 대한 요구사항 Test 작성 및 구현

티켓에 대한 요구사항 Test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 중복되지 않는 6개의 수로 이루어져 있다

chanhalee committed:
refactor(Budget) validation 메서드 static 속성 제거

validation 메서드 static 속성 제거

chanhalee committed:
feat(WinningNumbers) 티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현

티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현
- 등수에 대한 Enum 클래스 Prize 작성
- Ball -> Number로 클래스 명 변경
- Number Set을 보유하는 Numbers 클래스 생성
- WinningNumbers, TicketNumbers가 Numbers를 상속 받도록 구현

chanhalee committed:
feat(PrizeCalculator) 수익률계산기 테스트 작성 및 구현

수익률계산기 테스트 작성 및 구현

chanhalee committed:
feat(LottoMachine) LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

chanhalee committed:
feat(LottoMachine) LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

chanhalee committed:
refactor MVC 패턴에 맞게 패키지 분리

chanhalee committed:
feat(LottoView) view 레이어 구현

view 레이어 구현

chanhalee committed:
feat(LottoController) Controller 구현

Controller 구현

iimpala committed:
refactor(dto) dto 패키지 분리

iimpala committed:
refactor(LottoNumber) 클래스 이름 변경

도메인의 의미를 좀 더 명확하게 나타낼 수 있도록 클래스 이름 변경
- Number -> LottoNumber

iimpala committed:
refactor(LottoTicket) 클래스명 변경

도메인의 의미를 좀 더 명확하게 나타낼 수 있도록 클래스 이름 변경
- TicketNumbers -> LottoTicket

iimpala committed:
refactor(WinningNumbers) WinningNumbers와 Numbers의 상속관계 제거

WinningNumbers가 Numbers를 상속받는 구조에서 WinningNumbers가 LottoTicket을 가지는 구조로 변경

iimpala committed:
refactor(LottoTicket) LottoTicket과 Numbers의 상속관계 제거

불필요한 상속관계를 제거하고 Numbers 클래스를 삭제

iimpala committed:
refactor(all) 상수 추출

도메인 객체들에서 상수 추출

Rok93 reviewed (changes requested):
동현님 안녕하세요. 
로또 미션을 함께하게된 김경록입니다. 🙇‍♂️

몇몇 피드백 남겨두었으니 확인 부탁드려요. 😃
질문 주신 부분과 TODO로 적어주신 부분에 대해서도 코멘트 남겨두었으니 같이 확인해서 반영해주시면 될 것 같습니다. 🙏

미션 진행하면서 어려운 부분이있으면 언제든지 DM 주세요. 🙂

iimpala committed:
refactor(LottoNumber) lottoNumberPool 제거

- 1 ~ 45까지의 LottoNumber를 캐싱하여 불칠요한 인스턴스를 생성하지 않도록 변경
- LottoNumber 생성자를 private로 변경
- LottoNumber에 getValue(), getValues(), valueOf() 정적 메소드 추가
- LottoTicket 생성방식 변경(numberPool 제거)

iimpala committed:
refactor(WinningLotto) 클래스면 변경

도메인간 포함관계와 성격을 고려하여 WinningNumbers -> WinningLotto로 클래스명 변경

iimpala committed:
refactor(LottoTicketSeller) 클래스명 변경 및 책임 분리

- 로또 판매와 결과확인 두가지 역할을 담당하던 LottoMachine에서 결과 확인 로직을 WinningLotto에 위임하여 한가지 책임만 담당하도록 변경
- 이에 따라 LottoMachine -> LottoTicketSeller로 클래스명 변경

iimpala committed:
refactor(Prize) getPrize 메소드 위치 변경

당첨 등수를 판별하는 로직을 Prize 클래스로 위임하여 객체의 응집도를 높임
- 메소드 이름 변경(getPrize -> evaluate)
- 테스트 코드 추가(PrizeTest)

iimpala committed:
refactor(LottoResult) LottoResult 관련 구조 리팩토링

- 도메인의 의미를 명확히 하기 위헤 클래스명 변경(Budget -> LottoPurchaseBudget, PrizeCalculator -> LottoResult)
- 수익률 계산 로직을 PrizeCalculator에서 LottoPurchaseBudget로 위임
- WinningLotto에서 단일 Ticket의 Prize를 반환하는 calculatePrize 메소드 작성
- 결과 집계를 WinningLotto에서 LottoResult로 위임
- 리팩토링과 클래스 추가로 인한 테스트 추가 및 변경

iimpala committed:
refactor(dto) 도메인에서 toDto 메소드 분리

view 레이어에서 도메인의 Prize를 알고있는 문제를 해결하기 위해 controller에서 도메인과 dto를 변환하는 작업을 수행
- Prize에서 toString 오버라이딩
- dto 패키지 위치 이동
- dto에서 Prize 의존성 제거

iimpala committed:
refactor(exception) 에외 타입 변경

validation 실패시 던지는 예외 타입을 RuntimeException -> IllegalArgumentException으로 변경

Rok93 reviewed (approved):
동현님 피드백 반영 잘해주셨습니다. 👍
1단계는 이정도면 충분한 것 같아서 이만 머지하겠습니다. 👏👏

추가적으로 남겨둔 코멘트있는데 다음 단계 진행할 때, 같이 반영해주시면 될 것 같습니다. 🙂
다음 단계에서 뵐게요! 😃
